### PR TITLE
tap: fix orientation and corner cells

### PR DESCRIPTION
**Pull request by eder-matheus**

### CODE REVIEW

File: src/tap/src/tapcell.tcl [hunk 1]
```diff
@@ -672,18 +683,19 @@ proc insert_at_top_bottom {db rows masters endcap_master prefix} {
...
   return $topbottom_cnt
 }
 
-proc insert_at_top_bottom_helper {block top_bottom ori x_start x_end lly tap_nwintie_master tap_nwin2_master tap_nwin3_master tap_nwouttie_master tap_nwout2_master tap_nwout3_master prefix} {
+proc insert_at_top_bottom_helper {block top_bottom is_macro ori x_start x_end lly tap_nwintie_master tap_nwin2_master tap_nwin3_master tap_nwouttie_master tap_nwout2_master tap_nwout3_master prefix} {
   if {$top_bottom == 1} {
     # top
-    if { $ori == "MX" } {
+    if {($ori == "R0" && $is_macro) || \
+        ($ori == "MX" && !$is_macro)} {
```
Comment (maliberty) [inline src/tap/src/tapcell.tcl:698]: Would you explain this further - why are these particular orientations special?  Why does is_macro matter?

File: src/tap/src/tapcell.tcl [hunk 2]
```diff
@@ -584,6 +584,13 @@ proc insert_tapcells {db rows tapcell_master dist prefix} {
         set overlap [check_if_filled $x $tap_width $ori $row_fill_check]
         if {$overlap == 0} {
           build_cell $block $master $ori $x $lly $prefix
+        } elseif {[llength $overlap] == 2} {
```
Comment (maliberty) [inline src/tap/src/tapcell.tcl:587]: What scenario is being handled here?

File: src/tap/src/tapcell.tcl (hunk 1 above)
Comment (eder-matheus) [reply src/tap/src/tapcell.tcl:698]: tapcell is looking at row orientation here. After checking with Calibre, we've noticed that the master depended on where we're placing these top/bottom cells: at the top/bottom of the core area, or at the top/bottom of macros. With this update, Calibre reports 0 DRCs regarding tapcells.

Comment (maliberty) [reply src/tap/src/tapcell.tcl:698]: How does is_macro relate to row orientation?

Comment (maliberty) [reply src/tap/src/tapcell.tcl:698]: I want to make sure this is a general fix and not just for a specific issue.

Comment (eder-matheus) [reply src/tap/src/tapcell.tcl:698]: The cell we choose for placement at the top/bottom of rows and at the top/bottom of the core depends on the row orientation. For the core, we choose _IN_ cells when the orientation is MX, but for macros, we do it when the orientation is R0. In summary, the behavior of the top/bottom insertion depends if it is near macros or at the core.
I'm available to talk if it is not clear.

**Review by maliberty — APPROVED** [review]

**Pull request merged**